android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch

```diff
@@ -0,0 +1,1112 @@
++{
++	clear_bit(BTUSB_ISOC_RX_START, &data->flags);
++}
++
++static int bt_pcm_capture_trigger(struct snd_pcm_substream *substream,int cmd)
++{
++	struct btusb_data *data;
++	int err = 0;
++
++	data = snd_pcm_substream_chip(substream);
++	switch (cmd) {
++		case SNDRV_PCM_TRIGGER_START:
++		case SNDRV_PCM_TRIGGER_RESUME:
++			pr_emerg("%s - command: %s", __func__,
++				(cmd == SNDRV_PCM_TRIGGER_START) ?
++				"SNDRV_PCM_TRIGGER_START" : "SNDRV_PCM_TRIGGER_RESUME");
++			set_bit(BTUSB_ISOC_RX_START, &data->flags);
++			err = btusb_isoc_submit_rx_urbs(data);
++			if (err < 0)
++				pr_emerg("%s err:%d\n", __func__, err);
++			break;
++		case SNDRV_PCM_TRIGGER_STOP:
++		case SNDRV_PCM_TRIGGER_SUSPEND:
++			pr_emerg("%s - command: %s", __func__,
++				(cmd == SNDRV_PCM_TRIGGER_STOP) ?
++			  "SNDRV_PCM_TRIGGER_STOP" : "SNDRV_PCM_TRIGGER_SUSPEND");
++			btusb_isoc_stop_rx(data);
++			break;
++		default:
++			err = -EINVAL;
++	}
++	return err;
++}
++
++
++/* trigger is atomic can not used mutex sleep  or schedule another func */
++static int bt_pcm_playback_trigger(struct snd_pcm_substream *substream, int cmd)
++{
++	struct btusb_data *data = snd_pcm_substream_chip(substream);
++	int err = 0;
++
++	switch (cmd) {
++		case SNDRV_PCM_TRIGGER_START:
++		case SNDRV_PCM_TRIGGER_RESUME:
++			pr_emerg("%s - command: %s", __func__,
++				(cmd == SNDRV_PCM_TRIGGER_START)
++				? "SNDRV_PCM_TRIGGER_START" : "SNDRV_PCM_TRIGGER_RESUME");
++			set_bit(BTUSB_ISOC_TX_START, &data->flags);
++			btusb_isoc_submit_tx_urb(data);
++			break;
++		case SNDRV_PCM_TRIGGER_STOP:
++		case SNDRV_PCM_TRIGGER_SUSPEND:
++			//pr_emerg("%s - command: %s", __func__,
++			//	(cmd == SNDRV_PCM_TRIGGER_STOP)
++			//	? "SNDRV_PCM_TRIGGER_STOP" : "SNDRV_PCM_TRIGGER_SUSPEND");
++			btusb_isoc_stop_tx(data);
++			break;
++		default:
++			pr_emerg("%s - Invalid cmd: %d\n", __func__, cmd);
++			err = -EINVAL;
++	}
++	return err;
++}
++
++/* Atomic fuction This callback is called when the PCM middle layer inquires
++* the current hardware position on the buffer. The position must be returned
++* in frames, ranging from 0 to buffer_size - 1.*/
++static snd_pcm_uframes_t bt_pcm_pointer(struct snd_pcm_substream *substream)
++{
++	unsigned int playback_hwptr_done;
++	unsigned int capture_hwptr_done;
++	unsigned long flags;
++	struct btusb_data *data = snd_pcm_substream_chip(substream);
++
++	if (substream->stream == SNDRV_PCM_STREAM_CAPTURE) {
++		spin_lock_irqsave(&data->rxlock, flags);
++		capture_hwptr_done = data->capture_hwptr_done;
++		spin_unlock_irqrestore(&data->rxlock, flags);
++		return capture_hwptr_done /
++				(substream->runtime->frame_bits >> 3);
++	} else if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
++		spin_lock_irqsave(&data->txlock, flags);
++
++		playback_hwptr_done = data->playback_hwptr_done;
++
++		if (data->playback_transfer_done >=
++				(MAX_PACKET_SIZE - SCO_HDR_SIZE)) {
++			data->playback_transfer_done -=
++					(MAX_PACKET_SIZE - SCO_HDR_SIZE);
```

Comment:
Why this adjustment done here.. I believe it should be take care during processing itself.